OctoPrint/OctoPrint
# Fix/4493 - Defer M20 commands until initial capability report is done

## What does this PR do and why is it necessary?

This is a proposed fix for issue #4493.

This PR tracks the start and end of the first capability report and defers any `M20` commands until it's done.

The first `M20` command is queued before the response from `M115` is processed, meaning long filenames are not shown in the file list until it's refreshed.

## How was it tested? How can it be tested by the reviewer?

- Connect to any printer and observe the terminal output
- Verify no `M20` command is sent before the capability report is done
- Verify an `M20` command is sent shortly after the capability report finishes
  - with the `L` parameter if `EXTENDED_M20` is supported

## What are the relevant tickets if any?

This is a fix for issue #4493.

## Screenshots (if appropriate)

Not really screenshots but close enough.

<details>
  <summary>Terminal Output Before</summary>

  ```
    Connected to: VIRTUAL(read_timeout=10.0,write_timeout=10.0,options={'_config_version': 1, 'enabled': True}), starting monitor
    Send: N0 M110 N0*125
    Recv: start
    Send: N0 M110 N0*125
    Recv: Marlin: Virtual Marlin!
    Recv: \x80
    Recv: SD card ok
    Recv: ok
    Send: N0 M110 N0*125
    Changing monitoring state from "Connecting" to "Operational"
    Recv: ok
    Send: N0 M110 N0*125
    Recv: ok
    Send: N1 M115*39
    Recv: ok
    Send: N2 M20*19
    Recv: FIRMWARE_NAME:Virtual Marlin 1.0 PROTOCOL_VERSION:1.0
    Recv: Cap:AUTOREPORT_TEMP:1
    Recv: Cap:AUTOREPORT_SD_STATUS:1
    Recv: Cap:AUTOREPORT_POS:0
    Recv: Cap:EMERGENCY_PARSER:1
    Recv: Cap:EXTENDED_M20:1
    Recv: ok
    Send: M155 S2
    Recv: Begin file list
    Recv: End file list
    Recv: ok
    [...]
    Recv: ok
    [...]
    Recv: ok
```
</details>

<details>
  <summary>Terminal Output After</summary>

  ```
    Connected to: VIRTUAL(read_timeout=10.0,write_timeout=10.0,options={'_config_version': 1, 'enabled': True}), starting monitor
    Send: N0 M110 N0*125
    Recv: start
    Send: N0 M110 N0*125
    Recv: Marlin: Virtual Marlin!
    Recv: \x80
    Recv: SD card ok
    Recv: ok
    Send: N0 M110 N0*125
    Changing monitoring state from "Connecting" to "Operational"
    Recv: ok
    Recv: ok
    Send: N0 M110 N0*125
    Recv: ok
    Send: N1 M115*39
    Recv: FIRMWARE_NAME:Virtual Marlin 1.0 PROTOCOL_VERSION:1.0
    Recv: Cap:AUTOREPORT_TEMP:1
    Recv: Cap:AUTOREPORT_SD_STATUS:1
    Recv: Cap:AUTOREPORT_POS:0
    Recv: Cap:EMERGENCY_PARSER:1
    Recv: Cap:EXTENDED_M20:1
    Recv: ok
    Send: M155 S2
    Recv: ok
    [...]
    Recv: ok
    Send: M20 L
    Recv: Begin file list
    Recv: End file list
    Recv: ok
  ```
</details>

## Further notes

I'm unsure how likely/possible it is for the 'cap:' lines to be interrupted, but it could cause the M20 to be sent before the `EXTENDED_M20` capability is reported.

An alternative solution is to refresh the list as soon as the `EXTENDED_M20` capability is reported (like below), but that will double report the file list.

```python
elif (
          capability == self.CAPABILITY_EXTENDED_M20 and enabled
      ):
          self._logger.info(
              "Firmware states that it supports long filenames"
          )
         self.refreshSdFiles()
```

Thoughts are appreciated.
